v2/pkg/genruntime/extensions/postreconciliation_checker.go

```diff
@@ -0,0 +1,160 @@
+		severity: conditions.ConditionSeverityWarning,
+		reason:   conditions.ReasonPostReconcileFailure,
+		message:  reason,
+	}
+}
+
+func (r PostReconcileCheckResult) ReconciliationFailure() bool {
+	return r.action == postReconcileCheckResultTypeFailure
+}
+
+func (r PostReconcileCheckResult) ReconciliationSuccess() bool {
+	return r.action == postReconcileCheckResultTypeSuccess
+}
+
+func (r PostReconcileCheckResult) Message() string {
+	return r.message
+}
+
+// CreateConditionError returns an error that can be used to set a condition on the resource.
+func (r PostReconcileCheckResult) CreateConditionError() error {
+	return conditions.NewReadyConditionImpactingError(
+		errors.New(r.message),
+		r.severity,
+		r.reason)
+}
+
+// postReconcileCheckResultType is the type of result returned by PreReconcileCheck.
+type postReconcileCheckResultType string
+
+const (
+	postReconcileCheckResultTypeFailure postReconcileCheckResultType = "Failure"
+	postReconcileCheckResultTypeSuccess postReconcileCheckResultType = "Success"
+)
+
+// CreatePostReconciliationChecker creates a checker that can be used to check if we want to customise the condition on the resource after reconciliation.
+// If the resource in question has not implemented the PostReconciliationChecker interface, the provided default checker
+// is returned directly.
+// We also return a bool indicating whether the resource extension implements the PostReconciliationChecker interface.
+// host is a resource extension that may implement the PostReconciliationChecker interface.
+func CreatePostReconciliationChecker(
+	host genruntime.ResourceExtension,
+) (PostReconcileCheckFunc, bool) {
+	impl, ok := host.(PostReconciliationChecker)
+	if !ok {
+		return nil, false
+	}
+
+	return func(
+		ctx context.Context,
+		obj genruntime.MetaObject,
+		owner genruntime.MetaObject,
+		kubeClient kubeclient.Client,
+		armClient *genericarmclient.GenericClient,
+		log logr.Logger,
+	) (PostReconcileCheckResult, error) {
+		log.V(Status).Info("Extension post-reconcile check running")
+
+		result, err := impl.PostReconcileCheck(ctx, obj, owner, kubeClient, armClient, log, alwaysSucceed)
+		if err != nil {
+			log.V(Status).Info(
+				"Extension post-reconcile check failed",
+				"Error", err.Error())
+
+			// We choose to skip here so that things are definitely broken and the user will notice
+			// If we defaulted to always reconciling, the user might not notice that something is wrong
+			return PostReconcileCheckResultFailure("Extension PostReconcileCheck failed"), err
+		}
+
+		log.V(Status).Info("Extension post-reconcile check succeeded", "Result", result)
+
+		return result, nil
+	}, true
+}
+
+// alwaysSucceed is a PostReconciliationChecker that always indicates a reconciliation is successful.
+// We have this here, so we can set up a chain, even if it's only one link long.
+// When we start doing proper comparisons between Spec and Status, we'll have an actual chain of checkers.
```

Comment:
Will this extension be involved in comparisons between `Spec` & `Status`?

Reply:
good catch! I left this one behind. Removing it.